### ISSUE TITLE
Allow binary rather than unary decorators for smarts sampling

### ISSUE THREAD

**Issue by davidlmobley** [open-forcefield-group/smarty#13]
@jchodera , this one will require your attention.

We believe that to allow better exploration we need to allow binary rather than unary construction of SMARTS to sample. I haven't dug in too much to the details of the code on this aspect, but what @cbayly13  is telling us on this end is that he needs to be able to combine "pick a bond order" with "pick an element type or functional group" more easily - i.e. rather than taking a decorator that does both at once, being able to combine two decorators. In other words, instead of a unary decorator, he wants binary decorators (take this action such as a bond, apply it to that chemical group). 

So I think he wants to be able to revise this:

```
$(*~[#1])      hydrogen-adjacent
$(*~[#6])      carbon-adjacent
$(*~[#7])      nitrogen-adjacent
$(*~[#8])      oxygen-adjacent
$(*~[#9])      fluorine-adjacent
$(*~[#15])     phosphorous-adjacent
$(*~[#16])     sulfur-adjacent
$(*~[#17])     chlorine-adjacent
$(*~[#35])     bromine-adjacent
$(*~[#53])     iodine-adjacent
```

by replacing it with "pick a bond type" as one operation and "pick a functional group adjacent to it" as a second.

@cbayly13 , have I summarized properly? 

Ultimately, he will also want to optimize how these selections are made under the hood, i.e. trying a single bond basically works with everything, but double or triple bonds only make sense for particular elements/functional groups, etc. This may be an API issue.

**Comment by jchodera** [comment]
We could add a "binary decorators" category to our current "unary decorators" category, but this will significantly increase both the complexity and the rejection rate.

You can mimic this capability right now by manually expanding the list with sensible combinations. For example, you wouldn't want:

```
$(*-[#1])      hydrogen-adjacent-bondorder1
$(*=[#1])      hydrogen-adjacent-bondorder2
$(*#[#1])      hydrogen-adjacent-bondorder3
```

because most of those don't make sense, but you can try

```
$(*-[#1])      hydrogen-adjacent
$(*~[#6])      carbon-adjacent
$(*-[#6])      carbon-adjacent-bondorder1
$(*=[#6])      carbon-adjacent-bondorder2
$(*#[#6])      carbon-adjacent-bondorder3
$(*-[#7])      nitrogen-adjacent-bondorder1
$(*=[#7])      nitrogen-adjacent-bondorder2
$(*#[#7])      nitrogen-adjacent-bondorder3
$(*~[#8])      oxygen-adjacent
$(*-[#8])      oxygen-adjacent-bondorder1
$(*=[#8])      oxygen-adjacent-bondorder2
$(*~[#9])      fluorine-adjacent
$(*~[#15])     phosphorous-adjacent
$(*-[#15])     phosphorous-adjacent-bondorder1
$(*=[#15])     phosphorous-adjacent-bondorder2
$(*~[#16])     sulfur-adjacent
$(*-[#16])     sulfur-adjacent-bondorder1
$(*=[#16])     sulfur-adjacent-bondorder2
$(*-[#17])     chlorine-adjacent
$(*-[#35])     bromine-adjacent
$(*-[#53])     iodine-adjacent
```

which is essentially everything that is chemically reasonable.

If you still want the binary capability (perhaps because you have much grander plans than this example), let me know and I can implement it!

**Comment by davidlmobley** [comment]
@jchodera - Will discuss with @cbayly13 tomorrow, though I think he was fairly clear that he could already achieve this with the files now but has grander plans. 

Probably a better solution than you implement it is for you just to say, "if you wanted to implement that, you should do X and Y", or, if you're too busy, just say, "OK, I have no problem with trying that, but watch out for increased rejection rates!"

**Comment by jchodera** [comment]
> I think he was fairly clear that he could already achieve this with the files now but has grander plans.

OK, just wanted to clarify this!

This might be another good thing that you can try implementing.  The current child atomtype proposal code is [here](https://github.com/open-forcefield-group/smarty/blob/master/smarty/sampler.py#L280-L290).  You may want to factor this out into a function that will make it easier for you to propose new child atomtypes by different schemes.

To implement the binary scheme, you could add two more files, a `bond-type-decorators` and an `atom-type-decorators`, and then pick a bond type from one and an atom type from another.  You might even be able to come up with a more general strategy for constructing multi-component SMARTS strings, but you'd want to look over some more use cases first.

**Comment by davidlmobley** [comment]
Excellent, this is helpful. Thanks!

On Mon, Jun 20, 2016 at 5:38 PM, John Chodera notifications@github.com
wrote:

> I think he was fairly clear that he could already achieve this with the
> files now but has grander plans.
> 
> OK, just wanted to clarify this!
> 
> This might be another good thing that you can try implementing. The
> current child atomtype proposal code is here
> https://github.com/open-forcefield-group/smarty/blob/master/smarty/sampler.py#L280-L290.
> You may want to factor this out into a function that will make it easier
> for you to propose new child atomtypes by different schemes.
> 
> To implement the binary scheme, you could add two more files, a
> bond-type-decorators and an atom-type-decorators, and then pick a bond
> type from one and an atom type from another. You might even be able to come
> up with a more general strategy for constructing multi-component SMARTS
> strings, but you'd want to look over some more use cases first.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/open-forcefield-group/smarty/issues/13#issuecomment-227310232,
> or mute the thread
> https://github.com/notifications/unsubscribe/AGzUYcTZ7prOrkJlx_h_mumQKbiki00sks5qNzKNgaJpZM4I6QHB
> .

## 

David Mobley
dmobley@gmail.com
949-385-2436

**Comment by cbayly13** [comment]
Caught up on this thread. My main reason for binary decorators is because of the necessary additional complexity that John is worried about... I think we need it. For example, to distinguish an amide N ('N-C(=O)C') from a urea N ('N-C(=O)N') the decorator [#7&$(_~[#6])] will have to become not only  [#7X3&$(_~[#6(=O)])] for amide but [#7&$(_~[#6(=O)-NX3])] for urea. If I was "wizarding" this type, I would further generalize it to include thiourea with [#7&$(_~[#6(=O,S)-NX3])]. Special-casing every such context would be "wizarding" the decorators which defeats the purpose and beauty of the Bayesian approach. Should the wizard really lump together urea and thiourea? What is his rationale... nothing but chemical intuition from years of chemistry! Should the MCMC drive this? Ergo, binary decorators that begin with amine [#7X3&$(_~[#6])], discovers and keeps amide [#7X3&$(_~[#6(=O)])], and then maybe generalizes to amide and thioamide [#7&$(*~[#6(=O,S)])], etc.; I can't see this working any other way.

**Comment by jchodera** [comment]
Thanks for the clarification, @cbayly13

Out of curiosity, how does OpenEye SMARTS matching work when there is more than one atom in the query? We had talked about how to use the SMIRKS atom-association capability (and should maybe shift to this), but for a SMARTS query like `[#7&$(~[#6(=O,S)-NX3])]`, does it understand that this correctly types a single atom?

Also, my [decorators](https://github.com/open-forcefield-group/smarty/blob/master/examples/parm%40frosst/atomtypes/decorators.smarts#L2-L15) may have been ill-formed because they specify two atoms, like `$([*]=[*])`.

**Comment by cbayly13** [comment]
smarts above types a single atom. Equivalent smirks would be:
`[#7X3:1][#6](=O,S)[NX3]` , where the ':1' is the SMIRKS notation to tag that atom. The tagged atom would then get the single-atom parameters. For the N-C bond, SMIRKS would be `[#7X3:1][#6:2](=O,S)[NX3]` , the only difference being the ':2' ; bond angle N-C-N would SMIRKS as  `[#7X3:1][#6:2](=O,S)[NX3:3]`

**Comment by jchodera** [comment]
Note that surrounding the SMARTS/SMIRKS strings in single backticks ( ` ) will help make them much more readable!

**Comment by camizanette** [comment]
@cbayly13 and I discussed about the binary decorators yesterday. Christopher proposed we should separate the atoms from their bonds. So instead of having this:

```
$(*~[#1])      hydrogen-adjacent
$(*~[#6])      carbon-adjacent
```

We would have this:

```
$(*-z) simply bonded
$(*=z) doubly bonded
$(*#z) triply bonded
$(*:z) aromatic
$(*~[*]) any
```

Where 'z' gets replaced with a basetype atom.

The connectivity, total-h-count, formal charge, etc, will stay the same.
We want to make random SMARTS strings using the basetype atoms and our new decorators. And, after some analysis, Christopher said we need 2 or 3 decorators to cover most of the cases.
For example, to represent oxygen-hydrogen, we have:
`[#8X2+0$(*-[#1])]`

Thus, what was proposed is to generate randomly the new subtype where it can have 1, 2 or 3 decorators, randomly choose.
Example atom + one decorator:
`[#8X2] oxygen-dival`

It is good to have some insight about how randomly we can be in this case. We can have child having atom + decorator attached, or just go completely random.
This new binary decorators will remove the problem with atoms that are not in the data set trying to make new subtypes, since we will not have descriptor with atoms #21.

@cbayly13, please, add anything I may have forgot to mention.

**Comment by cbayly13** [comment]
Camila has given a good synopsis; the last point about how to introduce decorators on to the adjacent atom may require further thought, though Camila's starting point is interesting.  Her scheme is that the decorators are statically incorporated into the adjacent atom at child generation time, instead of being modified afterwards as an additional degree of freedom. I don't think we are actually reducing the formal degrees of freedom, but rather building it into the child generation. Let's see how it works; I think @bannanc should be included in the chemistry considerations.

**Comment by davidlmobley** [comment]
This was resolved by #48.
